File: 0112. 路径总和/DFS.swift
class Solution {
    var goal = 0

    func hasPathSum(_ root: TreeNode?, _ sum: Int) -> Bool {
        goal = sum
        return search(root)
    }

    func search(_ root: TreeNode?) -> Bool {
        if let root = root {
            if root.left == nil && root.right == nil {
                return root.val == goal
            } else {
                root.left?.val += root.val
                root.right?.val += root.val
                return search(root.left) || search(root.right)
            }
        } else {
            return false
        }
    }
}

File: 0538. 把二叉搜索树转换为累加树/DFS.js
// 144ms
var convertBST = function (root) {
    var sum = 0
    function travel(root) {
        if (root != null) {
            travel(root.right)
            sum += root.val
            root.val = sum
            travel(root.left)
        }
    }
    travel(root)
    return root
};